cptv/writer.pyx

```diff
@@ -0,0 +1,228 @@
+        fw.uint32(ord(Field.LAST_FFC_TIME), frame.last_ffc_time / timedelta(milliseconds=1))
+        fw.uint8(ord(Field.BIT_WIDTH), bit_width)
+        fw.uint32(ord(Field.FRAME_SIZE), len(frame_buf))
+        fw.write(ord(Section.FRAME), self.s)
+
+        frame_buf.tofile(self.s)
+
+    def close(self):
+        self.s.close()
+
+
+cdef class FieldWriter:
+
+    cdef int count
+    cdef object s
+
+    def __init__(self):
+        self.s = BytesIO()
+        self.count = 0
+
+    cpdef write(self, char section_type, object out):
+        out.write(struct.pack("<BB", section_type, self.count))
+        out.write(self.s.getbuffer())
+
+    cpdef timestamp(self, char code, object t):
+        cdef unsigned long long micros =  int(t.timestamp() * 1_000_000)
+        self.uint64(code, micros)
+
+    cpdef uint8(self, char code, unsigned char val):
+        self.s.write(struct.pack("<BBB", 1, code, val))
+        self.count += 1
+
+    cpdef uint32(self, char code, unsigned long val):
+        self.s.write(struct.pack("<BBL", 4, code, val))
+        self.count += 1
+
+    cpdef uint64(self, char code, unsigned long long val):
+        self.s.write(struct.pack("<BBQ", 8, code, val))
+        self.count += 1
+
+    cpdef float32(self, char code, float val):
+        self.s.write(struct.pack("<BBf", 4, code, val))
+        self.count += 1
+
+    cpdef string(self, char code, object val):
+        self.s.write(struct.pack("<BB", len(val), code))
+        self.s.write(val)
+        self.count += 1
+
+
+cdef class Compressor:
+
+    cdef int[:] frame_delta
+    cdef int[:] adj_delta
+    cdef unsigned short[:, :] prev_pix
+    cdef object out
+
+    def __init__(self):
+        self.frame_delta = np.zeros(ELEMS, dtype="int32")
+        self.adj_delta = np.zeros(ELEMS - 1, dtype="int32")
+        self.prev_pix = np.zeros((ROWS, COLS), dtype="uint16")
+        self.out = array("B")
+
+    @cython.boundscheck(False) # turn off bounds-checking for entire function
+    @cython.wraparound(False)  # turn off negative index wrapping for entire function
+    def next_frame(self, unsigned short[:, :] pix):
+        cdef int x, y, inc, i
+
+        # Generate the interframe delta.
+        # The output is written in a "snaked" fashion to avoid
+        # potentially greater deltas at the edges in the next stage.
+        for y in range(ROWS):
+            i = y * COLS
+            if y & 1 == 0:
+                inc = 1
+            else:
+                i += (COLS - 1)
+                inc = -1
+
+            for x in range(COLS):
+                self.frame_delta[i] = <long>(pix[y, x]) - <long>(self.prev_pix[y, x])
+                i += inc
+
+        self.prev_pix[...] = pix
+
+        # Now generate the adjacent "delta of deltas".
+        cdef unsigned long max_delta = 0
+        cdef unsigned long abs_delta
+        cdef long delta
+        for i in range(ELEMS - 1):
+            delta = self.frame_delta[i + 1] - self.frame_delta[i]
```

Comment:
Move adj delta calculation into frame_delta calculation loop

Reply:
I implemented this and benchmarked it with a large number of frames being written out. It made no measurably difference. My guess is the the removal of the extra loop is offset by the extra conditionals and housekeeping required do the adj_deltas calculations in the same loop. 

It made the code more complicated so I decided to stick with the current approach.